### PR TITLE
feat: add function for lint redocly config file

### PR DESCRIPTION
**Pull request by AlexVarchuk**
## What/Why/How?
Add function for lint .redocly config file.
In the near future, we should add the CLI command for it. And improve types.

## Check yourself

- [x] Code is linted
- [x] Tested with redoc/reference-docs/workflows
- [x] All new/updated code is covered with tests

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

File: packages/core/src/__tests__/lint.test.ts [hunk 1]
```diff
@@ -41,4 +41,52 @@ describe('lint', () => {
...
+      Array [
+        Object {
+          "location": Array [
+            Object {
+              "pointer": "#/",
+              "reportOnKey": true,
+              "source": "",
+            },
+          ],
+          "message": "The field \`apiDefinitions\` must be present on this level.",
```
Comment (RomanHotsiy) [inline packages/core/src/__tests__/lint.test.ts:84]: `apiDefinitions` is not required.

File: packages/core/src/lint.ts [hunk 2]
```diff
@@ -89,3 +91,46 @@ export async function lintDocument(opts: {
   });
   return ctx.problems.map((problem) => config.addProblemToIgnore(problem));
 }
+
+export async function LintDocumentConfig(opts: {
```
Comment (RomanHotsiy) [inline packages/core/src/lint.ts:95]: should start from lowercase, also, rename to simply `lintConfig`

File: packages/core/src/lint.ts [hunk 3]
```diff
@@ -89,3 +91,46 @@ export async function lintDocument(opts: {
...
+
+  // TODO: maybe need deprecate this part with preprocessors and regularRules.
+  const oasMajorVersion = openAPIMajor(oasVersion);
+  const rules = config.getRulesForOasVersion(oasMajorVersion);
+  const types = normalizeTypes(configTypes, config);
+  const preprocessors = initRules(rules as any, config, 'preprocessors', oasVersion);
+  const regularRules = initRules(rules as any, config, 'rules', oasVersion);
+  const normalizedVisitors = normalizeVisitors([...preprocessors, ...regularRules], types);
+
+  const resolvedRefMap = await resolveDocument({
```
Comment (RomanHotsiy) [inline packages/core/src/lint.ts:121]: I believe there is no need to resolve here, we dont' have $ref in config.

File: packages/core/src/types/redocly-yaml.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
+import { NodeType } from '.';
+
+const ConfigRoot: NodeType = {
+  properties: {
+    apiDefinitions: { type: 'object', additionalProperties: { type: 'string' } },
+    lint: 'Lint',
+    referenceDocs: 'ReferenceDocs',
+  },
+  required: ['apiDefinitions'],
```
Comment (RomanHotsiy) [inline packages/core/src/types/redocly-yaml.ts:9]: this is not required

File: packages/core/src/types/redocly-yaml.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    oas3_1Rules: { type: 'object'},
+    preprocessors: { type: 'object'},
+    oas2Preprocessors: { type: 'object'},
+    oas3_0Preprocessors: { type: 'object'},
+    oas3_1Preprocessors: { type: 'object'},
+    decorators: { type: 'object'},
+    oas2Decorators: { type: 'object'},
+    oas3_0Decorators: { type: 'object'},
+    oas3_1Decorators: { type: 'object'},
+    resolve: {
```
Comment (RomanHotsiy) [inline packages/core/src/types/redocly-yaml.ts:37]: Don't do deep types (deeper than 2 for arrays), it's not supported by the spec rule I believe (not 100% sure). Instead have named types.

File: packages/core/src/types/redocly-yaml.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+const ConfigRoot: NodeType = {
+  properties: {
+    apiDefinitions: { type: 'object', additionalProperties: { type: 'string' } },
+    lint: 'Lint',
+    referenceDocs: 'ReferenceDocs',
+  },
+  required: ['apiDefinitions'],
+};
+
+const Lint: NodeType = {
```
Comment (RomanHotsiy) [inline packages/core/src/types/redocly-yaml.ts:12]: I would name it as `ConfigLint`, same for rest

File: packages/core/src/lint.ts [hunk 7]
```diff
@@ -89,3 +91,46 @@ export async function lintDocument(opts: {
...
+    oasVersion: oasVersion,
+  };
+
+  // TODO: maybe need deprecate this part with preprocessors and regularRules.
+  const oasMajorVersion = openAPIMajor(oasVersion);
+  const rules = config.getRulesForOasVersion(oasMajorVersion);
+  const types = normalizeTypes(configTypes, config);
+  const preprocessors = initRules(rules as any, config, 'preprocessors', oasVersion);
+  const regularRules = initRules(rules as any, config, 'rules', oasVersion);
+  const normalizedVisitors = normalizeVisitors([...preprocessors, ...regularRules], types);
```
Comment (RomanHotsiy) [inline packages/core/src/lint.ts:119]: This is too verbose for simple case and also would be great to skip oasVersion hack.

Try this:

```suggestion
  // import specRule somewhere
  const rules = [ { severity: 'error', ruleId: 'spec', visitor: specRule({severity: 'error'}) } ];
  const normalizedVisitors = normalizeVisitors(rules, types);
```

Looks like you won't need config at all too.

**Review by RomanHotsiy — CHANGES_REQUESTED** [review]
Left a few comments.

File: packages/core/src/lint.ts (hunk 2 above)
Comment (AlexVarchuk) [reply packages/core/src/lint.ts]: I should be more accurate. Thanks

File: packages/core/src/lint.ts (hunk 7 above)
Comment (AlexVarchuk) [reply packages/core/src/lint.ts]: Thanks, it very helped me. I thought about it but couldn't implement it.

**Review by andriyl — APPROVED** [review]

**Review by RomanHotsiy — DISMISSED** [review]

**Review by RomanHotsiy — APPROVED** [review]

**Pull request merged**